File: This&that/_scripting_languages/Ruby/data_structures/graph/dijkstra.rb
# Dijkstra's algorithm

class Graph
    attr_accessor :size, :matrix

    INFINITY = Float::INFINITY

    def initialize(size)
        @size = size
        @matrix = []
        @dist = []
        @visited = []
        @nullValue = -1

        # initialize matrix
        size.times do |i|
            @matrix[i] = []
            size.times do |j|
                @matrix[i][j] = @nullValue
            end
        end
    end

    def connect(x, y, w)
        @matrix[x][y] = w
        @matrix[y][x] = w
    end

    def isConnected(x, y)
        @matrix[x][y] != @nullValue
    end

    #
    # Find the unvisited index in the dist array with the minimum distance value
    # this should be a minHeap...
    #
    def minDistance()
        currentMinValue = INFINITY
        indexWithMinDist = nil # indexWithMinDist = -100000 (old code)
        
        # Loop through the nodes
        @size.times do |i| 
            if  @visited[i] == false && 
                @dist[i] <= currentMinValue
                
                currentMinValue = @dist[i]
                indexWithMinDist = i
            end
        end
        return indexWithMinDist
    end

    #
    # print the distances from the nodes to the source
    #
    def printSolution()
        puts("Node Distance from Source\n")
        @size.times do |i|
            puts("#{i} \t\t #{@dist[i]}\n")
        end
    end

    #
    # Dijkstra's algorithm
    #
    def dijkstra(src)

        # Initialize @dist array and @visited array
        @size.times do |i| 
            @dist[i] = INFINITY # set all distances to INFINITY
            @visited[i] = false # indentify nodes as NOT @visited
        end

        # Distance of source node from itself is always 0
        @dist[src] = 0

        # Find shortest path for all nodes
        @size.times do
            # Pick node with the minimum distance to source, from the set of nodes not yet processed.
            # currentNode is always equal to src in first iteration.
            currentNode = minDistance()

            puts "CURRENT NODE: (i: #{currentNode}, dist: #{@dist[currentNode]}, visited: #{@visited[currentNode]}) \n"

            # Mark the picked node as processed
            @visited[currentNode] = true

            puts "CURRENT NODE: (i: #{currentNode}, dist: #{@dist[currentNode]}, visited: #{@visited[currentNode]}) \n\n"

            # Update dist value of the neighbor nodes of the currentNode.
            @size.times do |currentNeighbor|

                #
                # Update @dist[currentNeighbor] ONLY IF: 
                #
                # 1. the dist of currentNode is not INFINITY,
                # 2. there is an edge from currentNode to currentNeighbor (which means they are connected),
                # 3. currentNeighbor was NOT visited, 
                # 4. and total weight of path from src to currentNeighbor through currentNode is LESS 
                #    than current value of @dist[currentNeighbor]
                if  @dist[currentNode] != INFINITY &&
                    @matrix[currentNode][currentNeighbor] != @nullValue &&
                    @visited[currentNeighbor] == false &&
                    @dist[currentNode] + @matrix[currentNode][currentNeighbor] < @dist[currentNeighbor]

                    puts "\t\tCURRENT NEIGHBOR: (i: #{currentNeighbor}, dist: #{@dist[currentNeighbor]}, visited: #{@visited[currentNeighbor]}) \n"

                    @dist[currentNeighbor] = @dist[currentNode] + @matrix[currentNode][currentNeighbor]

                    puts "\t\tCURRENT NEIGHBOR: (i: #{currentNeighbor}, dist: #{@dist[currentNeighbor]}, visited: #{@visited[currentNeighbor]}) \n\n"
                end
            end
        end

        # print the constructed distance array
        printSolution()
    end

end

graph = Graph.new(3)

graph.connect(0, 1, 10)
graph.connect(0, 2, 100)
graph.connect(1, 2, 10)

graph.dijkstra(0)


#============================================================ OUTPUT:

# CURRENT NODE: (i: 0, dist: 0, visited: false) 
# CURRENT NODE: (i: 0, dist: 0, visited: true) 

#         CURRENT NEIGHBOR: (i: 1, dist: Infinity, visited: false) 
#         CURRENT NEIGHBOR: (i: 1, dist: 10, visited: false) 

#         CURRENT NEIGHBOR: (i: 2, dist: Infinity, visited: false) 
#         CURRENT NEIGHBOR: (i: 2, dist: 100, visited: false) 

# CURRENT NODE: (i: 1, dist: 10, visited: false) 
# CURRENT NODE: (i: 1, dist: 10, visited: true) 

#         CURRENT NEIGHBOR: (i: 2, dist: 100, visited: false) 
#         CURRENT NEIGHBOR: (i: 2, dist: 20, visited: false) 

# CURRENT NODE: (i: 2, dist: 20, visited: false) 
# CURRENT NODE: (i: 2, dist: 20, visited: true) 

# Node Distance from Source
# 0        0
# 1        10
# 2        20

File: This&that/_compiled_languages/C/data_structures/graph/dijkstra.c
//
// DIJKSTRA ALGORITHM
//

#include<stdio.h>
#include<stdlib.h>


#define INFINITY 100000
#define TRUE 1
#define FALSE 0
#define NOT_CONNECTED_VALUE -1


typedef struct Graph{
    int size;
    int ** array; // 2D array
} Graph;


Graph * createGraph(int size){
    Graph * graph = (Graph *)malloc(sizeof(Graph));
    graph->size = size;

    //
    // allocate space for 2D array
    //

    int i, j;

    graph->array = (int **)malloc(sizeof(int *) * size);

    for(i = 0; i < size; i++){
        graph->array[i] = (int *)malloc(sizeof(int) * size);
    }

    //
    // initialize 2D array
    //


    // Note that graph->array[i][j] is same as *(*(graph->array+i)+j)

    for(i = 0; i <  size; i++)
        for(j = 0; j < size; j++)
            graph->array[i][j] = NOT_CONNECTED_VALUE;  // OR *(*(graph->array+i)+j) = ++count

    // for(i = 0; i <  size; i++)
    //     for(j = 0; j < size; j++)
    //         printf("%d ", graph->array[i][j]);

    return graph;
}


void connect(Graph ** graph, int x, int y, int weight){
    (*graph)->array[x][y] = weight;
    (*graph)->array[y][x] = weight;
}

/**
 * isConnected is a function that returns:
 * 1 -> if connected
 * 0 -> if NOT connected
 */
int isConnected(Graph * graph, int x, int y){
    if( graph->array[x][y] == NOT_CONNECTED_VALUE ){
        puts("NO");
        return TRUE;
    }else{
        puts("YES");
        return FALSE;
    }
}

// void printGraph(Graph * graph){
//     graph->array[][];

//     char * row;
//     int i;
//     for(i = 0; i <  size; i++){
//         row
//     }

//     printf("");
// }


/**
 * This should have been a minHeap...
 */
int minDistance(Graph ** graph, int * dist, int * visited){

    int size = (*graph)->size;
    int currentMinValue = INFINITY;
    int indexWithMinDist = -1; // start with an invalid index!

    // Loop through the nodes
    int i;
    for(i = 0; i < size; i++){
        if(visited[i] == FALSE && dist[i] <= currentMinValue){
            currentMinValue = dist[i];
            indexWithMinDist = i;
        }
    }

    return indexWithMinDist;
}


void printSolution(Graph ** graph, int * dist){
    printf("Node Distance from Source\n");

    int size = (*graph)->size;
    int i;
    for(i = 0; i < size; i++){
        printf("%d \t\t %d\n", i, dist[i]);
    }
}

void dijkstra(Graph ** graph, int src){

    int size = (*graph)->size;
    int dist[size];
    int visited[size];

    // Initialize dist and visited arrays
    int i;
    for(i = 0; i < size; i++){
        dist[i] = INFINITY;
        visited[i] = FALSE;
    }

    // Distance of src from itself is always 0
    dist[src] = 0;

    for(i = 0; i < size; i++){
        int currentNode = minDistance(graph, dist, visited);

        printf("CURRENT NODE: (i: %d, dist: %d, visited: %d) \n", currentNode, dist[currentNode], visited[currentNode]);
        visited[currentNode] = TRUE;
        printf("CURRENT NODE: (i: %d, dist: %d, visited: %d) \n", currentNode, dist[currentNode], visited[currentNode]);

        int currentNeighbor;
        for(currentNeighbor = 0; currentNeighbor < size; currentNeighbor++){
            if(
                dist[currentNode] != INFINITY &&
                (*graph)->array[currentNode][currentNeighbor] != NOT_CONNECTED_VALUE &&
                visited[currentNeighbor] == FALSE &&
                dist[currentNode] + (*graph)->array[currentNode][currentNeighbor] < dist[currentNeighbor]
            ){
                printf("\t\tCURRENT NEIGHBOR: (i: %d, dist: %d, visited: %d) \n", currentNeighbor, dist[currentNeighbor], visited[currentNeighbor]);
                dist[currentNeighbor] = dist[currentNode] + (*graph)->array[currentNode][currentNeighbor];
                printf("\t\tCURRENT NEIGHBOR: (i: %d, dist: %d, visited: %d) \n", currentNeighbor, dist[currentNeighbor], visited[currentNeighbor]);
            }
        }
    }

    printSolution(graph, dist);
}


int main(){

    Graph * graph = createGraph(3);

    connect(&graph, 0, 1, 10);
    connect(&graph, 0, 2, 100);
    connect(&graph, 1, 2, 10);

    // isConnected(graph, 1, 1);
    // isConnected(graph, 2, 2);
    // isConnected(graph, 3, 3);
    // isConnected(graph, 4, 4);

    dijkstra(&graph, 0);

    return 0;
}


/* OUTPUT:

CURRENT NODE: (i: 0, dist: 0, visited: 0)
CURRENT NODE: (i: 0, dist: 0, visited: 1)
        CURRENT NEIGHBOR: (i: 1, dist: 100000, visited: 0)
        CURRENT NEIGHBOR: (i: 1, dist: 10, visited: 0)
        CURRENT NEIGHBOR: (i: 2, dist: 100000, visited: 0)
        CURRENT NEIGHBOR: (i: 2, dist: 100, visited: 0)
CURRENT NODE: (i: 1, dist: 10, visited: 0)
CURRENT NODE: (i: 1, dist: 10, visited: 1)
        CURRENT NEIGHBOR: (i: 2, dist: 100, visited: 0)
        CURRENT NEIGHBOR: (i: 2, dist: 20, visited: 0)
CURRENT NODE: (i: 2, dist: 20, visited: 0)
CURRENT NODE: (i: 2, dist: 20, visited: 1)
Node Distance from Source
0        0
1        10
2        20

*/
